### DISCUSSION TITLE
Cannot update formulae since Python needs Apple Command Lines Tools installed (which it is)

### DISCUSSION THREAD

**Discussion by vladoski** [Homebrew/discussions#5255] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.16
ORIGIN: https://github.com/Homebrew/brew
HEAD: bd1155be8f50998429a795c15a69c8fe75250510
Last commit: 18 hours ago
Core tap JSON: 02 Apr 08:45 UTC
Core cask tap JSON: 02 Apr 08:45 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: /usr/bin/nano
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.3.1-arm64
CLT: N/A
Xcode: 15.2
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  python-click

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  lua@5.3
  youtube-dl
```

### Description of issue

Hi guys, everytime that I need to update something that has python as a dependency (e.g. ffmpeg), it won't let me since it gives me an error:
```
Error: python@3.12: the bottle needs the Apple Command Line Tools to be installed.
  You can install them, if desired, with:
    xcode-select --install
```

I have Apple Command Line Tools installed: `xcode-select --version` outputs `xcode-select version 2405.` I'm using pyenv to manage my python versions, I've tried updating brew formulae with both the system version (shimmed by pyenv but located here `/opt/homebrew/bin/python3`) and with versions installed via pyenv. Might pyenv be the problem?  

Also `brew list | rg python` outputs:
```
python-click
python@3.10
python@3.11
python@3.12
python@3.9
```  

Thanks for helping! :)

**Comment by SMillerDev** [comment]
What does `xcode-select -p` output?

**Reply by vladoski** [reply]
`/Applications/Xcode.app/Contents/Developer`

**Reply by SMillerDev** [reply, marked as answer]
That means you don't have the commandline tools, you have Xcode tools for the command line. They're not the same. What happens if you run `xcode-select --install`?

**Reply by vladoski** [reply]
you're right I thought that Xcode tools and Apple Command Tools were the same thing, my bad.
Thanks!

**Reply by naderghanbari** [reply]
@vladoski I wouldn't really call it your bad! Those two things are the same thing for all/most intents and purposes! They may be called different things, but what matters is they provide the same toolchain (Xcode provides more than Command Line Tools). That's why you can switch between the two using the `--switch flag`. Otherwise why would this command exist in the first place?

```shell
> sudo xcode-select --switch
--help        -h  -- display help information                                                                                   
--install         -- install command-line developer tools                                                                       
--print-path  -p  -- display path of active developer directory                                                                 
--reset       -r  -- reset to default developer directory                                                                       
--switch      -s  -- specify path of active developer directory                                                                 
--version     -v  -- display version information      
```

Shall we invoke the Liskov substitution principle? 😉

Mine is currently pointing to the Xcode location at 

```shell
> xcode-select --print-path 
/Applications/Xcode.app/Contents/Developer
```
and everything works fine (including building and making), but homebrew doesn't work. It still insists on having the Comand Line Tools installed, even though I beleive Xcode tools are a super set of Command Line Tools and include more. 

There is a myriad of questions on SO, Reddit, and all over the place about this. Apple could have done a much better job here. When there are so many confusions, one can't blame the community. Their deisgn is wrong imho. If it were up to me, I would centralize the common parts of the toolchain, and if Xcode needs more, it can augment it (instead of replacing it), then there would be one source of truth for the toolchain.
